numpy/ma/extras.py

```diff
@@ -708,34 +708,42 @@ def _median(a, axis=None, out=None, overwrite_input=False):
             asorted = a
     else:
         asorted = sort(a, axis=axis)
+
     if axis is None:
         axis = 0
     elif axis < 0:
         axis += a.ndim
+    if axis < 0 or axis >= a.ndim:
+        raise ValueError("Axis value out of bounds")
 
     if asorted.ndim == 1:
         idx, odd = divmod(count(asorted), 2)
-        return asorted[idx - (not odd) : idx + 1].mean()
+        return asorted[idx + odd - 1 : idx + 1].mean()
 
-    counts = asorted.shape[axis] - (asorted.mask).sum(axis=axis)
+    if asorted.mask is nomask:
+        counts = asorted.shape[axis]
```

Comment:
How about the `count` function? I note that it also handles a tuple of axes, which median currently does not.
```
    def count(self, axis=None, keepdims=np._NoValue):
        """
        Count the non-masked elements of the array along the given axis.
```
The count function is already imported and used for the 1-d case in `_median`.